File: function/nodejs_14_x/src/error/error_response.ts
import { AWSError } from 'aws-sdk';
import { Parser } from 'xml2js';
import ErrorCode from './error_code';
import { BaseObjectContext, GetObjectContext } from '../s3objectlambda_event.types';
import S3 from 'aws-sdk/clients/s3';
import { PromiseResult } from 'aws-sdk/lib/request';
import { Response, Headers } from 'node-fetch';
import { IErrorResponse } from '../s3objectlambda_response.types';

/**
 * Generates a response to Amazon S3 Object Lambda when there is an error
 * with a GET_OBJECT request.
 */
export async function getErrorResponse (s3Client: S3, requestContext: GetObjectContext,
  errorCode: ErrorCode, errorMessage: string, headers: Headers = new Headers()): Promise<PromiseResult<{}, AWSError>> {
  console.log(`Returning an error [${errorCode}] ${errorMessage} to the Object Lambda Access Point`);

  return s3Client.writeGetObjectResponse({
    RequestRoute: requestContext.outputRoute,
    RequestToken: requestContext.outputToken,
    StatusCode: ERROR_TO_STATUS_CODE_MAP[errorCode],
    ErrorCode: errorCode,
    ErrorMessage: errorMessage,
    ...headers
  }).promise();
}

/**
 * Generates a response to Amazon S3 Object Lambda when there is an error
 * with a HEAD_OBJECT / LIST_OBJECTS_V1 / LIST_OBJECTS_V2 request.
 */
export function errorResponse (requestContext: BaseObjectContext,
  errorCode: ErrorCode, errorMessage: string, headers: Map<string, object> = new Map<string, object>()): IErrorResponse {
  console.log(`Returning an error [${errorCode}] ${errorMessage} to the Object Lambda Access Point`);

  return {
    statusCode: ERROR_TO_STATUS_CODE_MAP[errorCode],
    errorCode: errorCode,
    errorMessage: errorMessage,
    ...Object.fromEntries(headers)
  };
}

/**
 * Sends back the error that was received from S3 on a GET_OBJECT request
 */
export async function getResponseForS3Errors (s3Client: S3, requestContext: GetObjectContext, objectResponse: Response,
  headers: Headers, objectResponseBody: Buffer): Promise<PromiseResult<{}, AWSError>> {
  const objectResponseData = await new Parser().parseStringPromise(objectResponseBody);
  console.log(`Encountered an S3 Error, status code: ${objectResponse.status}. Forwarding this to the Object Lambda Access Point.`);

  return s3Client.writeGetObjectResponse({
    RequestRoute: requestContext.outputRoute,
    RequestToken: requestContext.outputToken,
    StatusCode: objectResponse.status,
    ErrorCode: objectResponseData.Code,
    ErrorMessage: `Received ${objectResponse.statusText} from the supporting Access Point.`,
    ...headers
  }).promise();
}

/**
 * Sends back the error that was received from S3 on a HEAD_OBJECT / LIST_OBJECTS request
 */
export function responseForS3Errors (objectResponse: Response): IErrorResponse {
  console.log(`Encountered an S3 Error, status code: ${objectResponse.status}. Forwarding this to the Object Lambda Access Point.`);

  return {
    statusCode: objectResponse.status,
    errorMessage: `Received ${objectResponse.statusText} from the supporting Access Point.`
  };
}

/**
 * Maps error codes to HTTP Status codes based on the {@link https://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html | Error Responses}
 * mapping in Amazon S3.
 */
const ERROR_TO_STATUS_CODE_MAP = {
  [ErrorCode.INVALID_REQUEST]: 400,
  [ErrorCode.INVALID_PART]: 400,
  [ErrorCode.INVALID_RANGE]: 416,
  [ErrorCode.NO_SUCH_KEY]: 404
};

File: function/python_3_9/src/error/error_response.py
from xml.etree import ElementTree


def write_error_response_for_s3(s3_client, request_context, object_response):
    """
    write WriteGetObjectResponse to the S3 client for error msg received from s3
    :param s3_client: s3 client
    :param request_context: Requests that was sent to supporting Access Point
    :param object_response: Response received from supporting Access Point
    :return: WriteGetObjectResponse
    """
    root = ElementTree.fromstring(object_response.content.decode('utf-8'))
    error_code = root.find('Code').text
    return s3_client.write_get_object_response(
        RequestRoute=request_context['outputRoute'],
        RequestToken=request_context['outputToken'],
        StatusCode=object_response.status_code,
        ErrorCode=error_code,
        ErrorMessage='Received {} from the supporting Access Point.'.format(error_code),
    )


def write_error_response(s3_client, request_context, status_code, error_code, error_message):
    """
    write WriteGetObjectResponse to the S3 client for AWS Lambda errors
    :param s3_client: s3 client
    :param request_context: Requests that was sent to supporting Access Point
    :param status_code: Http status code for the type of error
    :param error_code: Error Code
    :param error_message: Error Message
    :return: WriteGetObjectResponse
    """
    return s3_client.write_get_object_response(
        RequestRoute=request_context['outputRoute'],
        RequestToken=request_context['outputToken'],
        StatusCode=status_code,
        ErrorCode=error_code,
        ErrorMessage=error_message
    )
